core/src/replay_stage.rs

```diff
@@ -2642,60 +2651,116 @@ impl ReplayStage {
         last_vote_refresh_time: &mut LastVoteRefreshTime,
         voting_sender: &Sender<VoteOp>,
         wait_to_vote_slot: Option<Slot>,
-    ) {
-        let last_voted_slot = tower.last_voted_slot();
-        if last_voted_slot.is_none() {
-            return;
-        }
+    ) -> bool {
+        let Some(heaviest_bank_on_same_fork) = heaviest_bank_on_same_fork.as_ref() else {
+            return false;
+        };
+        let Some(latest_landed_vote_slot) =
+            progress.my_latest_landed_vote(heaviest_bank_on_same_fork.slot())
+        else {
+            return false;
+        };
+        let Some(last_voted_slot) = tower.last_voted_slot() else {
+            return false;
+        };
+        let Some(last_vote_block_height) = progress
+            .get_fork_stats(latest_landed_vote_slot)
+            .map(|fs| fs.block_height)
+        else {
+            return false;
+        };
 
-        // Refresh the vote if our latest vote hasn't landed, and the recent blockhash of the
-        // last attempt at a vote transaction has expired
-        let last_voted_slot = last_voted_slot.unwrap();
-        if my_latest_landed_vote > last_voted_slot
+        // If our last landed vote on this fork is greater than the vote recorded in our tower
+        // this means that our tower is old AND on chain adoption has failed. Warn the operator
+        // as they could be submitting slashable votes.
+        if latest_landed_vote_slot > last_voted_slot
             && last_vote_refresh_time.last_print_time.elapsed().as_secs() >= 1
         {
             last_vote_refresh_time.last_print_time = Instant::now();
-            info!(
+            warn!(
                 "Last landed vote for slot {} in bank {} is greater than the current last vote \
-                 for slot: {} tracked by Tower",
-                my_latest_landed_vote,
+                 for slot: {} tracked by tower. This indicates a bug in the on chain adoption logic",
+                latest_landed_vote_slot,
                 heaviest_bank_on_same_fork.slot(),
                 last_voted_slot
             );
+            datapoint_error!(
+                "adoption_failure",
+                ("latest_landed_vote_slot", latest_landed_vote_slot, i64),
+                (
+                    "heaviest_bank_on_fork",
+                    heaviest_bank_on_same_fork.slot(),
+                    i64
+                ),
+                ("last_voted_slot", last_voted_slot, i64)
+            );
+        }
+
+        if latest_landed_vote_slot >= last_voted_slot {
+            // Our vote or a subsequent vote landed do not refresh
+            return false;
+        }
+        if last_vote_block_height.saturating_add(REFRESH_VOTE_BLOCKHEIGHT as u64)
+            > heaviest_bank_on_same_fork.block_height()
```

Comment:
Do we have to check block height instead of blockhash queue? Or is this just design choice because the tx isn't actually expired anymore and `.is_hash_valid_for_age(&last_vote_tx_blockhash.unwrap(), 16)` feels hacky?

Reply:
I could go either way with this, technically we don't actually need to refresh the block hash since it's not expired but we should for completeness. I'll just reuse the `is_hash_valid_for_age` since it's convenient

Reply:
I guess the nice (or maybe terrible in some niche cases...) thing about not refreshing the block hash is that the tx will get de-duped when appropriate. Seems possible we could land the same vote twice now, right?